openproblems/data/immune_cells.py

```diff
@@ -0,0 +1,56 @@
+from . import utils
+
+import os
+import scanpy as sc
+import scprep
+import tempfile
+
+URL = "https://ndownloader.figshare.com/files/25717328"
+
+
+@utils.loader
+def load_immune(test=False):
+    """Download immune human data from figshare."""
+    if test:
+        # load full data first, cached if available
+        adata = load_immune(test=False)
+
+        # Subsample pancreas data
+        adata = adata[:, :500].copy()
+        utils.filter_genes_cells(adata)
+
+        # We are using the final annotation column from the original file
+        # for the cell type annotation, the batch column contains the name
+        # of the study the data originates from
+
+        keep_final_annotations = adata.obs["final_annotation"].dtype.categories[[0, 3]]
+        keep_batchs = adata.obs["batch"].dtype.categories[[0, -3, -2]]
+        keep_batch_idx = adata.obs["batch"].isin(keep_batchs)
+        keep_final_annotation_idx = adata.obs["final_annotation"].isin(
+            keep_final_annotations
+        )
+        adata = adata[keep_batch_idx & keep_final_annotation_idx].copy()
+
+        sc.pp.subsample(adata, n_obs=500)
+        # Note: could also use 200-500 HVGs rather than 200 random genes
```

Comment:
I don't understand the test subsampling here entirely. First you subsample to the first 500 genes, then filter cells and genes, then you filter batches, filter cells again randomly and then filter genes and cells again. Why do this twice? And why subsample to the first 500 genes first? Why not just subsample to a fixed number of cells and genes (not via subsample, but deterministically), and then filter cells and genes once?

Reply:
I copied that from the pancreas loader without checking, probably makes sense to just remove the additional filtering

Reply:
Your first filtering is only on genes... that doesn't make too much sense for me either. Maybe immediately filter on genes and cells in the beginning, then do the rest?